From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: sclp: handle zero-length event buffers
References: bnc#477666,LTC#51641

Symptom:     Support Element (SE) restart may cause Linux on LPAR to hang.
Problem:     During SE restart, some SE versions may under certain conditions
             present a malformed Read Event Data response block to Linux which
             causes an endless loop in function sclp_dispatch_evbufs.
Solution:    Stop event dispatching loop when a zero-length event buffer was
             recognized.

Acked-by: John Jolly <jjolly@suse.de>

---
 drivers/s390/char/sclp.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

Index: linux-sles11/drivers/s390/char/sclp.c
===================================================================
--- linux-sles11.orig/drivers/s390/char/sclp.c
+++ linux-sles11/drivers/s390/char/sclp.c
@@ -280,8 +280,11 @@ sclp_dispatch_evbufs(struct sccb_header 
 	rc = 0;
 	for (offset = sizeof(struct sccb_header); offset < sccb->length;
 	     offset += evbuf->length) {
-		/* Search for event handler */
 		evbuf = (struct evbuf_header *) ((addr_t) sccb + offset);
+		/* Check for malformed hardware response */
+		if (evbuf->length == 0)
+			break;
+		/* Search for event handler */
 		reg = NULL;
 		list_for_each(l, &sclp_reg_list) {
 			reg = list_entry(l, struct sclp_register, list);
